jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildAndCacheApplicationLayerStep.java

```diff
@@ -0,0 +1,132 @@
+      Cache cache) {
+    try (Timer ignored = new Timer(buildConfiguration.getBuildLogger(), DESCRIPTION)) {
+      return ImmutableList.of(
+          new BuildAndCacheApplicationLayerStep(
+              "dependencies",
+              listeningExecutorService,
+              buildConfiguration,
+              sourceFilesConfiguration.getDependenciesFiles(),
+              sourceFilesConfiguration.getDependenciesPathOnImage(),
+              cache),
+          new BuildAndCacheApplicationLayerStep(
+              "resources",
+              listeningExecutorService,
+              buildConfiguration,
+              sourceFilesConfiguration.getResourcesFiles(),
+              sourceFilesConfiguration.getResourcesPathOnImage(),
+              cache),
+          new BuildAndCacheApplicationLayerStep(
+              "classes",
+              listeningExecutorService,
+              buildConfiguration,
+              sourceFilesConfiguration.getClassesFiles(),
+              sourceFilesConfiguration.getClassesPathOnImage(),
+              cache));
+    }
+  }
+
+  private final String layerType;
+  private final BuildConfiguration buildConfiguration;
+  private final List<Path> sourceFiles;
+  private final String extractionPath;
+  private final Cache cache;
+
+  private final ListeningExecutorService listeningExecutorService;
+  @Nullable private ListenableFuture<CachedLayer> listenableFuture;
+
+  private BuildAndCacheApplicationLayerStep(
+      String layerType,
+      ListeningExecutorService listeningExecutorService,
+      BuildConfiguration buildConfiguration,
+      List<Path> sourceFiles,
+      String extractionPath,
+      Cache cache) {
+    this.layerType = layerType;
+    this.listeningExecutorService = listeningExecutorService;
+    this.buildConfiguration = buildConfiguration;
+    this.sourceFiles = sourceFiles;
+    this.extractionPath = extractionPath;
+    this.cache = cache;
+  }
+
+  @Override
+  public ListenableFuture<CachedLayer> getFuture() {
+    if (listenableFuture == null) {
+      listenableFuture = listeningExecutorService.submit(this);
+    }
+    return listenableFuture;
+  }
+
+  @Override
+  public CachedLayer call()
+      throws IOException, LayerPropertyNotFoundException, CacheMetadataCorruptedException {
+    String description = "Building " + layerType + " layer";
+
+    buildConfiguration.getBuildLogger().lifecycle(description + "...");
+
+    try (Timer ignored = new Timer(buildConfiguration.getBuildLogger(), description)) {
+      // Don't build the layer if it exists already.
+      CachedLayer cachedLayer = new CacheReader(cache).getUpToDateLayerBySourceFiles(sourceFiles);
+      if (cachedLayer != null) {
+        return cachedLayer;
+      }
+
+      ReproducibleLayerBuilder reproducibleLayerBuilder =
+          new ReproducibleLayerBuilder(sourceFiles, extractionPath);
+
+      cachedLayer = new CacheWriter(cache).writeLayer(reproducibleLayerBuilder);
+      // TODO: Remove
```

Comment:
Any reason we haven't removed this yet?

Reply:
I think we'll actually keep this in for now since it is a useful debug message to see which application layer was which digest.